### PR TITLE
Bug 1543003, Added CNS Upgrade Content

### PR DESCRIPTION
**Pull request by ahardin-rh**
https://bugzilla.redhat.com/show_bug.cgi?id=1543003

http://file.rdu.redhat.com/%7Eahardin/03012018/upgrade-cns/upgrading/automated_upgrades.html#special-considerations-for-glusterfs

### CODE REVIEW

File: upgrading/cns_upgrades.adoc [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+. Upgrade standard nodes running application containers.
++
+[NOTE]
+====
+If any of the nodes in those groups are running GlusterFS , perform step 4 of
+this procedure at the same time. GlusterFS nodes must be upgraded along with
+other nodes in their class (`app` versus `infra`), one at a time.
+====
+
+. Upgrade the CNS cluster one node at a time, commenting out other nodes.
```
Comment (ahardin-rh) [inline upgrading/cns_upgrades.adoc]: @mrobson @jarrpa What commands should be listed for this procedure? I would like to be as specific as possible.

**Comment by ahardin-rh** [comment]
@mrobson @jarrpa This is my first pass. PTAL. Thanks!

**Comment by ahardin-rh** [comment]
@tmorriso-rh FYI

File: upgrading/cns_upgrades.adoc (hunk 1 above)
Comment (jarrpa) [reply upgrading/cns_upgrades.adoc]: What do you mean by "commenting out other nodes"?

**Review by jarrpa — CHANGES_REQUESTED** [review]
HI there!

One big thing upfront is that, for the most part, you can't and must not reference CNS. There is no such thing as CNS nodes, nor CNS pods, and certainly no such thing as CNS in the upstream. Most mentions of CNS can simply be replaced with "GlusterFS", though information pertaining specifically to containerized GlusterFS should say so.

Note also that most of this needs to be backported for 3.6 as well.

I don't have access to an OpenShift environment right now to get you the exact commands you need. That will probably have to wait until tomorrow at least.

File: upgrading/cns_upgrades.adoc (hunk 1 above)
Comment (ahardin-rh) [reply upgrading/cns_upgrades.adoc]: This was pulled in from a comment from @mrobson here: https://bugzilla.redhat.com/show_bug.cgi?id=1543003#c2

**Comment by ahardin-rh** [comment]
@jarrpa Thanks for your feedback and for clarifying the CNS usage. I will do some rework. 

Once we get that content finalized, we can identify what needs to go into 3.6 docs. I can introduce that in a separate PR if necessary.

**Comment by ahardin-rh** [comment]
@jarrpa I made some updates. PTAL. Thanks again!

File: upgrading/manual_upgrades.adoc [hunk 2]
```diff
@@ -31,6 +31,10 @@ Ensure that you have met all
 xref:../install_config/install/prerequisites.adoc#install-config-install-prerequisites[prerequisites]
 before proceeding with an upgrade. Failure to do so can result in a failed
 upgrade.
+
+If you are using CNS, see
+xref:../upgrading/cns_upgrades.adoc#upgrading-CNS[Upgrading CNS on
+{product-title}] before proceeding.
```
Comment (jarrpa) [inline upgrading/manual_upgrades.adoc]: Still have references to CNS.

File: upgrading/glusterfs_upgrades.adoc [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+
+.. Run the upgrade playbook.
+
+.. Relabel the node.
+
+.. Wait for the gluster pod to respawn and appear.
+
+.. `rsh` into the pod and verify all volumes are healed
+
+. Repeat step 4 for for GlusterFS nodes 2 and 3.
```
Comment (jarrpa) [inline upgrading/glusterfs_upgrades.adoc]: You can probably omit this line, since you already basically say this in line 73.

**Review by jarrpa — CHANGES_REQUESTED** [review]
I'm not sure I understand the point of the first section of `glusterfs_upgrades.adoc`. Specifically the list of commands. Items 1 and 2 list commands that are not really related to the descriptions they're providing. I'd suggest removing everything above line 37.

Also, it looks like there are some unrelated changes to DNS that have made it into this PR. This makes it a little harder for me to properly review your changes so please move them to their own PR is possible.

**Comment by ahardin-rh** [comment]
@jarrpa Thanks. I made another round of changes. 
@mrobson Please review, as the content was primarily sourced from your comments in the BZ. 
Regarding comments to DNS, I am certainly not a subject matter on this and I'm not sure what would be considered DNS vs CNS, so perhaps Matt can provide further guidance there.

**Review by jarrpa — CHANGES_REQUESTED** [review]
DNS is Domain Name Service, CNS is Container-Native Storage. Basically, just remove all changes from `manual_upgrades.adoc`. :)

**Comment by ahardin-rh** [comment]
@jarrpa Oh, sorry I was confused. I see what you mean. The only thing added to that file was the link to the Upgrading GlusterFS on OCP topic. I think when the content was wrapped, it pushed the formatting a bit, but there are actually no other changes to that page.

**Comment by jarrpa** [comment]
@ahardin-rh Okay. I still want them out of this PR. ;)

**Review by mrobson — COMMENTED** [review]
First review based on customer's successful usage of this method in OCP 3.7

**Comment by ahardin-rh** [comment]
@mrobson @jarrpa The latest changes are complete and the preview build is updated. PTAL and let me know if further changes are needed.

@jarrpa I apologize, but I am unable to get rid of the extra highlighting after formatting and reformatting. I think it has something to do with my editor. The only thing changed in that file is the added link, however.

File: _topic_map.yml [hunk 4]
```diff
@@ -509,6 +509,8 @@ Topics:
   File: migrating_etcd
 - Name: Downgrading
   File: downgrade
+- Name: Upgrading When Using GlusterFS Container Native Storage
```
Comment (jarrpa) [inline _topic_map.yml]: Do not use `GlusterFS Container Native Storage` to refer to GlusterFS. Please use the phrase `Containerized GlusterFS`.

File: upgrading/automated_upgrades.adoc [hunk 5]
```diff
@@ -42,6 +42,11 @@ Ensure that you have met all
 xref:../install_config/install/prerequisites.adoc#install-config-install-prerequisites[prerequisites]
 before proceeding with an upgrade. Failure to do so can result in a failed
 upgrade.
+
+If you are using GlusterFS, see
+xref:../upgrading/glusterfs_upgrades#upgrading-openshift-when-using-glusterfs[Upgrading
+{product-title} When Using GlusterFS Container Native Storage] before
```
Comment (jarrpa) [inline upgrading/automated_upgrades.adoc]: Do not use `GlusterFS Container Native Storage` to refer to GlusterFS. Please use the phrase `Containerized GlusterFS`.

File: upgrading/glusterfs_upgrades.adoc [hunk 6]
```diff
@@ -0,0 +1,100 @@
+[[upgrading-openshift-when-using-glusterfs]]
+= Upgrading {product-title} When Using GlusterFS Container Native Storage
```
Comment (jarrpa) [inline upgrading/glusterfs_upgrades.adoc]: Do not use `GlusterFS Container Native Storage` to refer to GlusterFS. Please use the phrase `Containerized GlusterFS`.

File: upgrading/glusterfs_upgrades.adoc [hunk 7]
```diff
@@ -0,0 +1,100 @@
...
+====
+If any of the nodes in those groups are running GlusterFS , perform step 4 of
+this procedure at the same time. GlusterFS nodes must be upgraded along with
+other nodes in their class (`app` versus `infra`), one at a time.
+====
+
+. Upgrade the {product-title} nodes running GlusterFS one at a time.
+
+.. Run `oc get daemonset` to verify the label found under `NODE-SELECTOR` is:
++
```
Comment (jarrpa) [inline upgrading/glusterfs_upgrades.adoc]: Remove `is` and replace with ``. The default value is `storagenode=glusterfs`.``

File: upgrading/glusterfs_upgrades.adoc [hunk 8]
```diff
@@ -0,0 +1,100 @@
...
+----
+Default: storagenode=glusterfs
+----
```
Comment (jarrpa) [inline upgrading/glusterfs_upgrades.adoc]: Remove the above and merge it with the bullet point.

File: upgrading/glusterfs_upgrades.adoc [hunk 9]
```diff
@@ -0,0 +1,100 @@
...
++
+----
+$ oc label node <node_name> storagenode-
+----
++
+This will cause the GlusterFS pod to terminate on that node.
+
+.. Run the upgrade playbook on the single node where you terminated GlusterFS.
+
+.. When the upgrade completes, relabel the node to be what it was previously:
```
Comment (jarrpa) [inline upgrading/glusterfs_upgrades.adoc]: `relabel the node with the DaemonSet selector:`

File: upgrading/glusterfs_upgrades.adoc [hunk 10]
```diff
@@ -0,0 +1,100 @@
...
++
+----
+$ oc label node <node_name> <daemonset_label>
+----
+
+.. Wait for the gluster pod to respawn and appear.
```
Comment (jarrpa) [inline upgrading/glusterfs_upgrades.adoc]: `GlusterFS`

File: upgrading/glusterfs_upgrades.adoc [hunk 11]
```diff
@@ -0,0 +1,100 @@
...
+
+.. `rsh` into the pod and verify all volumes are healed:
```
Comment (jarrpa) [inline upgrading/glusterfs_upgrades.adoc]: `oc rsh`

File: upgrading/glusterfs_upgrades.adoc [hunk 12]
```diff
@@ -0,0 +1,100 @@
...
++
+----
+$ oc rsh <gluster_pod_name>
```
Comment (jarrpa) [inline upgrading/glusterfs_upgrades.adoc]: Add the following line:
```
$ for vol in `gluster volume list`; do gluster volume heal $vol info; done
```

File: upgrading/glusterfs_upgrades.adoc [hunk 13]
```diff
@@ -0,0 +1,100 @@
...
++
+----
+$ gluster volume list
+----
+
+.. Run the following command on each volume to check the self-heal status:
++
+----
+$ gluster volume heal info
+----
```
Comment (jarrpa) [inline upgrading/glusterfs_upgrades.adoc]: Remove the above two bullet points and replace them with the suggested script above.

File: upgrading/glusterfs_upgrades.adoc [hunk 14]
```diff
@@ -0,0 +1,100 @@
...
++
+Ensure all of the volumes are healed and there are no outstanding tasks.
```
Comment (jarrpa) [inline upgrading/glusterfs_upgrades.adoc]: ``The heal info command lists all pending entries for a given volume's heal process. A volume is considered healed when `Number of entries` for that volume is 0.``

**Review by jarrpa — CHANGES_REQUESTED** [review]
Additional comments made.

@ahardin-rh What editor are you using?

**Comment by ahardin-rh** [comment]
@jarrpa Thanks! Updates complete. 
To answer your question, I use Atom as my editor

QE reviewed and approved.

File: upgrading/glusterfs_upgrades.adoc [hunk 15]
```diff
@@ -0,0 +1,87 @@
...
+be given for GlusterFS to complete all of its healing operations before
+GlusterFS on the next node is terminated. This could leave the cluster in a bad
+or unknown state. Therefore, the following procedure is recommended.
+
+[[upgrading-openshift-when-running-glusterfs]]
+== Upgrading {product-title} When Using Containerized GlusterFS
+
+. Upgrade the control plane (the master nodes and etcd nodes).
+
+. Upgrade standard infra nodes (router, registry, logging and metrics).
```
Comment (adellape) [inline upgrading/glusterfs_upgrades.adoc]: `infra`?

Comment (adellape) [reply upgrading/glusterfs_upgrades.adoc]: "logging, and metrics"?

**Pull request merged**